drivers/photo-electronic/adpd188/adpd188.c

```diff
@@ -0,0 +1,871 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdlib.h>
+#include "adpd188.h"
+#include "error.h"
+#include "delay.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Initialize the ADPD188 driver.
+ * @param device - The ADPD188 descriptor.
+ * @param init_param - The structure that contains the ADPD188 initialization
+ *                     parameters.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd188_init(struct adpd188_dev **device,
+		     struct adpd188_init_param *init_param)
+{
+	int32_t ret;
+	struct adpd188_dev *dev;
+	uint16_t reg_data;
+
+	dev = (struct adpd188_dev *)calloc(1, sizeof (*dev));
+	dev->phy_opt = init_param->phy_opt;
+
+	if(dev->phy_opt == ADPD188_SPI)
+		ret = spi_init((struct spi_desc **)&dev->phy_desc,
+			       (const struct spi_init_param *)&init_param->phy_init);
+	else if(dev->phy_opt == ADPD188_I2C)
+		ret = i2c_init((struct i2c_desc **)&dev->phy_desc,
+			       (const struct i2c_init_param *)&init_param->phy_init);
+	else
+		ret = FAILURE;
+	if(ret != SUCCESS)
+		goto error_dev;
+
+	ret = adpd188_reg_read(dev, ADPD188_REG_I2CS_ID, &reg_data);
+	if(ret != SUCCESS)
+		goto error_phy;
+	if(reg_data != 0x00C8)
+		goto error_phy;
+
+	ret = adpd188_sw_reset(dev);
+	if(ret != SUCCESS)
+		goto error_phy;
+
+	ret = gpio_get(&dev->gpio0, &init_param->gpio0_init);
+	if(ret != SUCCESS)
+		goto error_phy;
+	ret = gpio_get(&dev->gpio1, &init_param->gpio1_init);
+	if(ret != SUCCESS)
+		goto error_gpio0;
+
+	ret = gpio_direction_input(dev->gpio0);
+	if(ret != SUCCESS)
+		goto error_gpio1;
+	ret = gpio_direction_input(dev->gpio1);
+	if(ret != SUCCESS)
+		goto error_gpio1;
+
+	*device = dev;
+
+	return SUCCESS;
+
+error_gpio1:
+	gpio_remove(dev->gpio1);
+error_gpio0:
+	gpio_remove(dev->gpio0);
+error_phy:
+	if(dev->phy_opt == ADPD188_SPI)
+		spi_remove(dev->phy_desc);
+	else if(dev->phy_opt == ADPD188_I2C)
+		i2c_remove(dev->phy_desc);
+error_dev:
+	free(dev);
+
+	return FAILURE;
+}
+
+/**
+ * @brief Free resources allocated by adpd188_init().
+ * @param dev - The ADPD188 descriptor.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t adpd188_remove(struct adpd188_dev *dev)
+{
+	int32_t ret;
+
+	if(dev->phy_opt == ADPD188_SPI)
+		ret = spi_remove(dev->phy_desc);
+	else if(dev->phy_opt == ADPD188_I2C)
+		ret = i2c_remove(dev->phy_desc);
+	else
+		ret = FAILURE;
+	if(ret != SUCCESS)
+		return FAILURE;
```

Comment:
suggestion: maybe add an `error_dev` label as in the init function and perform an unconditional jump instead of multiple `return FAILURE;` statements.

Reply:
Interesting suggestion, but what would be the benefit? The detriment I see is that in case of error there would be two Program Counter loads instead of one.